### PR TITLE
Fix mask 101 GPU support in GpuCAReduce

### PR DESCRIPTION
**Pull request by adbrebs**
#2254 @abergeron @bouthilx

### CODE REVIEW

**Comment by adbrebs** [comment]
Resulting graph:

```python
>>> import theano
>>> import numpy as np

>>> x = theano.shared(np.ones((2,3,4,5), dtype=theano.config.floatX))
>>> f = theano.function([], theano.tensor.sum(x, (0, 1, 3)))
>>> theano.printing.debugprint(f, print_type=True)
HostFromGpu [id A] <TensorType(float32, vector)> ''   9
 |GpuCAReduce{add}{1,0,1,1} [id B] <CudaNdarrayType(float32, vector)> ''   8
   |GpuDimShuffle{0,1,2,x} [id C] <CudaNdarrayType(float32, (False, False, False, True))> ''   7
     |GpuReshape{3} [id D] <CudaNdarrayType(float32, 3D)> ''   6
       |<CudaNdarrayType(float32, 4D)> [id E] <CudaNdarrayType(float32, 4D)>
       |MakeVector{dtype='int64'} [id F] <TensorType(int64, vector)> ''   5
         |Elemwise{Mul}[(0, 0)] [id G] <TensorType(int64, scalar)> ''   4
         | |Shape_i{0} [id H] <TensorType(int64, scalar)> ''   3
         | | |<CudaNdarrayType(float32, 4D)> [id E] <CudaNdarrayType(float32, 4D)>
         | |Shape_i{1} [id I] <TensorType(int64, scalar)> ''   2
         |   |<CudaNdarrayType(float32, 4D)> [id E] <CudaNdarrayType(float32, 4D)>
         |Shape_i{2} [id J] <TensorType(int64, scalar)> ''   1
         | |<CudaNdarrayType(float32, 4D)> [id E] <CudaNdarrayType(float32, 4D)>
         |Shape_i{3} [id K] <TensorType(int64, scalar)> ''   0
           |<CudaNdarrayType(float32, 4D)> [id E] <CudaNdarrayType(float32, 4D)>
```

File: theano/sandbox/cuda/opt.py [hunk 1]
```diff
@@ -862,20 +862,31 @@ def local_gpu_careduce(node):
...
+                            raise Exception('Reduction mask [1, 0, 1, 1] is'
+                                            'supposed to be supported.')
+
+                    rval = host_from_gpu(
+                        new_greduce(gpu_new_x))
+
+                    # Restore the expected shape of the output
+                    if rval.ndim != out.ndim:
+                        rval = rval.reshape(
+                            tensor.stack(shape_of[out]))
```
Comment (abergeron) [inline theano/sandbox/cuda/opt.py]: Is this really required?

Comment (abergeron) [inline theano/sandbox/cuda/opt.py]: Sorry I just was that it was moved from above.

**Comment by abergeron** [comment]
This looks all good to me.  Only thing that could make it better would be a test maybe.

**Comment by nouiz** [comment]
For the tests, you can probably just uncomment two line in the current tests. uncomment the case 101 and 1101 (or 1011, I forgot which one)

**Comment by adbrebs** [comment]
I can't find the corresponding tests, could you please show me where they are?

**Comment by abergeron** [comment]
https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/tests/test_basic_ops.py#L116
https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/tests/test_basic_ops.py#L136
https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/tests/test_basic_ops.py#L142

**Comment by adbrebs** [comment]
Thanks @abergeron

**Comment by adbrebs** [comment]
@abergeron I have checked and the tests run well on GPU.

**Pull request merged**